### PR TITLE
Verify commands & tools for snapshot data

### PR DESCRIPTION
**Pull request by ethan-kr**
## Proposed changes

- This PR is derived from https://github.com/ethereum/go-ethereum/pull/21724
- Added snapshot subcommand in order to verify snapshot data
- It creates a trie based on snapshot data and compare the created root hash with the root hash in the header
- Added snapshot subcommand in order to trace statetrie nodes
- It reads all the nodes in the statetrie, looking for any missing nodes.
- Added iteration tools for snapshot nodes & statetrie nodes

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

#1229

### CODE REVIEW

File: tests/snapshot_iterator.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package main
```
Comment (hqjang-pepper) [inline tests/snapshot_iterator.go:1]: You should add copyrights on the top of the code.

File: tests/statetrie_iterator.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
+package main
```
Comment (hqjang-pepper) [inline tests/statetrie_iterator.go:1]: You should add copyrights on the top of the code.

File: tests/snapshot_iterator.go (hunk 1 above)
Comment (hqjang-pepper) [inline tests/snapshot_iterator.go:1]: I think you have to change package name (main -> test)

File: tests/snapshot_iterator.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+func main() {
```
Comment (hqjang-pepper) [inline tests/snapshot_iterator.go:10]: Is it okay if there is main() function here?

File: tests/snapshot_iterator.go (hunk 1 above)
Comment (ethan-kr) [reply tests/snapshot_iterator.go]: Sorry some late.
This code has been changed to verify-tools.

Comment (ethan-kr) [reply tests/snapshot_iterator.go]: Sorry some late.
This code has been changed to verify-tools.

File: tests/snapshot_iterator.go (hunk 3 above)
Comment (ethan-kr) [reply tests/snapshot_iterator.go]: Sorry some late.
This code has been changed to verify-tools.

File: tests/statetrie_iterator.go (hunk 2 above)
Comment (ethan-kr) [reply tests/statetrie_iterator.go]: Sorry some late.
This code has been changed to verify-tools.

File: blockchain/state/statedb.go [hunk 4]
```diff
@@ -1092,3 +1102,48 @@ func (s *StateDB) GetContractStorageRoot(contractAddr common.Address) (common.Ha
 	}
 	return contract.GetStorageRoot(), nil
 }
+
+func Tracer(db Database, root common.Hash) (resultErr error) {
+	once.Do(func() {
```
Comment (jeongkyun-oh) [inline blockchain/state/statedb.go:1107]: Why `once.Do` method is used here? Is this necessary?

File: blockchain/state/statedb.go [hunk 5]
```diff
@@ -1092,3 +1102,48 @@ func (s *StateDB) GetContractStorageRoot(contractAddr common.Address) (common.Ha
...
+		go func() {
```
Comment (jeongkyun-oh) [inline blockchain/state/statedb.go:1108]: When does this goroutine is terminated?

File: cmd/utils/nodecmd/snapshot.go [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+func iterateTrie(ctx *cli.Context) error {
+	stack := MakeFullNode(ctx)
+	dbm := stack.OpenDatabase(getConfig(ctx))
+	sdb, err := state.New(common.Hash{}, state.NewDatabase(dbm), nil)
+	if err != nil {
+		return fmt.Errorf("Failed to open newDB trie : %v", err)
+	}
+
+	logger.Info("TrieDB Iterator Start", "node count : all node count, nil node count : key or value is nil node count")
+	Cnt, nilCnt := uint64(0), uint64(0)
```
Comment (jeongkyun-oh) [inline cmd/utils/nodecmd/snapshot.go:254]: `Cnt` can be private.

File: cmd/utils/nodecmd/snapshot.go [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+	go func() {
```
Comment (jeongkyun-oh) [inline cmd/utils/nodecmd/snapshot.go:255]: Shouldn't we handle goroutine lifecycle?

File: blockchain/state/statedb.go [hunk 8]
```diff
@@ -1092,3 +1102,48 @@ func (s *StateDB) GetContractStorageRoot(contractAddr common.Address) (common.Ha
 	}
 	return contract.GetStorageRoot(), nil
 }
+
+func Tracer(db Database, root common.Hash) (resultErr error) {
```
Comment (jeongkyun-oh) [inline blockchain/state/statedb.go:1106]: How about changing this method name to `TraverseTrie`?

File: cmd/utils/nodecmd/snapshot.go [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+
+	children, err := tridb.NodeChildren(root)
+	if err != nil {
+		return fmt.Errorf("Fail get childrens of root : %v", err)
+	}
+
+	var wait sync.WaitGroup
+	wait.Add(len(children))
+	for _, child := range children {
+		go func(eth common.Hash) {
```
Comment (jeongkyun-oh) [inline cmd/utils/nodecmd/snapshot.go:234]: Any purpose of `eth` name? It seems like `ethereum` :P

File: cmd/utils/nodecmd/snapshot.go [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+		return errors.New("empty root")
+	}
+
+	logger.Info("Trace Start", "BlockNum", blockNumber)
+
+	sdb, err := state.New(root, state.NewDatabase(dbm), nil)
+	if err != nil {
+		return fmt.Errorf("Failed to open newDB trie : %v", err)
+	}
+	tridb := sdb.Database().TrieDB()
```
Comment (henry-will) [inline cmd/utils/nodecmd/snapshot.go:235]: Is triedb or trieDB better than tridb?

File: cmd/utils/nodecmd/snapshot.go [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+				utils.DynamoDBRegionFlag,
+				utils.DynamoDBIsProvisionedFlag,
+				utils.DynamoDBReadCapacityFlag,
+				utils.DynamoDBWriteCapacityFlag,
+				utils.LevelDBCompressionTypeFlag,
+				utils.DataDirFlag,
+			},
+			Description: `
+klaytn statedb iterate-triedb
+Coount the number of nodes in the state-trie db.
```
Comment (henry-will) [inline cmd/utils/nodecmd/snapshot.go:112]: Is it a typo in Count?

File: cmd/utils/nodecmd/snapshot.go [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+	}()
+
+	it := sdb.Database().TrieDB().DiskDB().GetStateTrieDB().NewIterator(nil, nil)
+	for it.Next() {
+		Cnt++
+		if it.Key() == nil || it.Value() == nil || bytes.Equal(it.Key(), []byte("")) || bytes.Equal(it.Value(), []byte("")) {
+			nilCnt++
+		}
+	}
+	it.Release()
```
Comment (henry-will) [inline cmd/utils/nodecmd/snapshot.go:327]: In other code references, Iterator uses defer Release(). How about moving the 'defer it.Release()' statement to appear right after the 'it' variable is assigned.

**Review by henry-will — CHANGES_REQUESTED** [review]

**Pull request closed without merging**